src/difacto/async_sgd.h

```diff
@@ -294,54 +254,213 @@ struct AdaGradHandle : public ISGDHandle {
       w[i] -= eta * grad;
     }
   }
-};
+};//end AdaGradHandle
+
+struct FTRLEntry{
+    FTRLEntry() {}
+    ~FTRLEntry() {Clear();}
+    inline void Clear(){
+  	if(size > 1) {delete [] w; delete [] sqc_grad;}
+	size = 0; w = NULL; sqc_grad = NULL;
+    }
+    inline void Resize(int n) {
+        if (n < size) { size = n; return; }
+
+        float* new_w = new float[n];
+        float* new_cg = new float[n+1];
+
+        if (size == 1) {
+            new_w[0] = w_0(); new_cg[0] = sqc_grad_0(); new_cg[1] = z_0();
+        } else {
+             memcpy(new_w, w, size * sizeof(float));
+             memcpy(new_cg, sqc_grad, (size+1) * sizeof(float));
+             Clear();
+        }
+        w = new_w; sqc_grad = new_cg; size = n;
+    }
+    inline float& w_0() { return size == 1 ? *(float *)&w : w[0]; }
+    inline float w_0() const { return size == 1 ? *(float *)&w : w[0]; }
+    inline float& sqc_grad_0() {
+	return size == 1 ? *(float *)&sqc_grad : sqc_grad[0];
+    }
+    inline float& z_0() {
+	return size == 1 ? *(((float *)&sqc_grad)+1) : sqc_grad[1];
+    }
+    inline void Load(Stream *fi){
+	fi->Read(&size, sizeof(size)) ;
+        if (size == 1) {
+            fi->Read(&w, sizeof(float*));
+            fi->Read(&sqc_grad, sizeof(float*));
+        } else {
+      	    w = new float[size];
+            sqc_grad = new float[size+1];
+            fi->Read(w, sizeof(float)*size);
+            fi->Read(sqc_grad, sizeof(float)*(size+1));
+            ISGDHandle::new_V += size - 1;
+        }
+        if (w_0() != 0) ++ ISGDHandle::new_w;
+    }
+    inline void Save(Stream *fo) const {
+ 	fo->Write(&size, sizeof(size));
+        if (size == 1) {
+            fo->Write(&w, sizeof(float*));
+            fo->Write(&sqc_grad, sizeof(float*));
+        } else {
+            fo->Write(w, sizeof(float)*size);
+            fo->Write(sqc_grad, sizeof(float)*(size+1));
+        }
+    }
+    inline bool Empty() const {return (w == 0 && size == 1); }
+
+    unsigned fea_cnt = 0;
+    int size = 1;
+    float *w = 0;
+    float *sqc_grad = NULL;
+};//end FTRLEntry
+
+struct FTRLHandle : public ISGDHandle{
```

Comment:
there're a lot of repetition between FTRLEntry/FTRLHandle and AdaGradEntry/AdaGradHandle,  just change the updateV()?